### PR TITLE
feat(model): add condition to Model.delete

### PR DESCRIPTION
**Pull request by tmartensen**
### Summary:

Add ability to add a ConditionExpression to a delete request via the `condition` setting on Model.delete. Add method to clear empty objects from DynamoDb requests to avoid DynamoDB error that occurs when `ExpressionAttributeValues` object is empty.

#### Schema
```
const UserSchema = new dynamoose.Schema({"id": Number, "name": String});
```

#### Model
```
const User = dynamoose.model("User", UserSchema);
```

#### General
```
const condition = new dynamoose.Condition().filter("id").exists();
await User.delete(1, {condition});
```

### GitHub linked issue:

#---

### Type (select 1):
- [ ] Bug fix
- [X] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement

- [ ] Test added to report bug (GitHub issue #--- @---)
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [X] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [X] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [X] Yes
- [ ] No

### Other:
- [X] I have read through and followed the Contributing Guidelines
- [X] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [X] I have updated the Dynamoose documentation (if required) given the changes I made
- [X] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [X] I have ensured the following commands are successful from the root of the project directory
  - [X] `npm test`
  - [X] `npm run lint`
- [X] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/master/LICENSE)
- [X] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [X] I have filled out all fields above

### CODE REVIEW

File: docs/docs/guide/Model.md [hunk 1]
```diff
@@ -512,7 +512,8 @@ You can also pass in an object for the optional `settings` parameter that is an
 
 | Name | Description | Type | Default |
 |------|-------------|------|---------|
-| return | What the function should return. Can be null, or `request`. In the event this is set to `request` the request Dynamoose will make to DynamoDB will be returned, and no request to DynamoDB will be made. If this is `request`, the function will not be async anymore. | String \| null | null |
+| return | What the function should return. Can be null, or `request`. In the event this is set to `request` the request Dynamoose will make to DynamoDB will be returned, and no request to DynamoDB will be made. If this is `request`, the function will not be async anymore. | String | null | null |
+| condition | This is an optional instance of a Condition for the delete. | [dynamoose.Condition](Condition.md) | null
```
Comment (fishcharlie) [inline docs/docs/guide/Model.md:516]:
```suggestion
| condition | This is an optional instance of a Condition for the delete. | [dynamoose.Condition](Condition) | null
```

File: docs/docs/guide/Model.md [hunk 2]
```diff
@@ -512,7 +512,8 @@ You can also pass in an object for the optional `settings` parameter that is an
 
 | Name | Description | Type | Default |
 |------|-------------|------|---------|
-| return | What the function should return. Can be null, or `request`. In the event this is set to `request` the request Dynamoose will make to DynamoDB will be returned, and no request to DynamoDB will be made. If this is `request`, the function will not be async anymore. | String \| null | null |
+| return | What the function should return. Can be null, or `request`. In the event this is set to `request` the request Dynamoose will make to DynamoDB will be returned, and no request to DynamoDB will be made. If this is `request`, the function will not be async anymore. | String | null | null |
```
Comment (fishcharlie) [inline docs/docs/guide/Model.md:515]: Why does this have any changes?

```suggestion
| return | What the function should return. Can be null, or `request`. In the event this is set to `request` the request Dynamoose will make to DynamoDB will be returned, and no request to DynamoDB will be made. If this is `request`, the function will not be async anymore. | String \| null | null |
```

File: lib/utils/object/index.ts [hunk 3]
```diff
@@ -5,6 +5,8 @@ import pick = require("./pick");
 import keys = require("./keys");
 import entries = require("./entries");
 import equals = require("./equals");
+import {clearEmpties} from "./clear_empties";
+
```
Comment (fishcharlie) [inline lib/utils/object/index.ts:9]:
```suggestion
```

File: lib/utils/object/clear_empties.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import {GeneralObjectOrValue} from "./types";
+
+export const clearEmpties = <T>(object: GeneralObjectOrValue<T>): GeneralObjectOrValue<T> => {
+	// TODO: implement deepClone for immutability of the passed in object?
```
Comment (fishcharlie) [inline lib/utils/object/clear_empties.ts:4]: I don't think any of the other `utils/object` methods do this, do they?

Comment (fishcharlie) [reply lib/utils/object/clear_empties.ts:4]:
```suggestion
```

File: lib/utils/object/clear_empties.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+	Object.keys(object).forEach((key) => {
```
Comment (fishcharlie) [inline lib/utils/object/clear_empties.ts:5]: Do we only want this to go one level deep?

I would rely on the other methods in `utils/object` to handle this. Such as `utils.object.keys`. That way it works for nested properties.

File: lib/utils/object/clear_empties.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+		if (typeof object[key] !== "object") {
+			return;
+		}
+		clearEmpties(object[key]);
```
Comment (fishcharlie) [inline lib/utils/object/clear_empties.ts:9]: With the comment above, we shouldn't need to do this.

File: test/unit/utils/object/clear_empties.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+		},
+		{
+			"input": {"id": 1, "name": "Bob", "address": {}},
+			"output": {"id": 1, "name": "Bob"}
+		},
+		{
+			"input": {"id": 1, "name": "Bob", "address": {"country": "world", "zip": 12345, "other": {}}},
+			"output": {"id": 1, "name": "Bob", "address": {"country": "world", "zip": 12345}}
+		}
+	];
```
Comment (fishcharlie) [inline test/unit/utils/object/clear_empties.js:30]: Would love to see some tests for empty arrays here.

**Review by fishcharlie — CHANGES_REQUESTED** [review]

File: lib/Condition.ts [hunk 8]
```diff
@@ -329,5 +329,6 @@ Condition.prototype.requestObject = function (this: Condition, settings: Conditi
 			return object;
 		}, {[settings.conditionString]: settings.conditionStringType === "array" ? [] : "", "ExpressionAttributeNames": {}, "ExpressionAttributeValues": {}});
 	}
-	return main(this.settings.conditions);
+
+	return utils.object.clearEmpties(main(this.settings.conditions));
```
Comment (fishcharlie) [inline lib/Condition.ts:333]: Also. How are we handling this in other cases? We might have to remove some code in other places we call this method to remove the empty objects.

**Review by fishcharlie — CHANGES_REQUESTED** [review]

File: lib/utils/object/clear_empties.ts (hunk 5 above)
Comment (tmartensen-chwy) [reply lib/utils/object/clear_empties.ts:5]: This method uses recursion to cycle through the keys. I tried replacing with `utils.object.keys` and it failed the nested tests.

Comment (tmartensen) [reply lib/utils/object/clear_empties.ts:5]: This method uses recursion to cycle through the keys. I tried replacing with utils.object.keys and it failed the nested tests.

File: test/unit/utils/object/clear_empties.js (hunk 7 above)
Comment (tmartensen-chwy) [reply test/unit/utils/object/clear_empties.js:38]: Added.

File: lib/utils/object/clear_empties.ts (hunk 4 above)
Comment (tmartensen-chwy) [reply lib/utils/object/clear_empties.ts]: I don't believe so. I'll leave this as is.

File: lib/utils/object/index.ts (hunk 3 above)
Comment (tmartensen-chwy) [reply lib/utils/object/index.ts:9]: Removed.

File: docs/docs/guide/Model.md (hunk 2 above)
Comment (tmartensen-chwy) [reply docs/docs/guide/Model.md]: Reverted.

File: docs/docs/guide/Model.md (hunk 1 above)
Comment (tmartensen-chwy) [reply docs/docs/guide/Model.md]: Changed.

File: test/unit/utils/object/clear_empties.js (hunk 7 above)
Comment (tmartensen) [reply test/unit/utils/object/clear_empties.js:38]: Done.

**Review by fishcharlie — APPROVED** [review]

**Comment by fishcharlie** [comment]
I messed up. Need @tmartensen to give me access to fix things, or to fix things so I can approve and merge.

Things still to do:

1. Fix lint errors
2. https://github.com/dynamoose/dynamoose/pull/1011#discussion_r502091487

File: test/unit/Model.js [hunk 9]
```diff
@@ -3702,7 +3702,7 @@ describe("Model", () => {
 					});
 				});
 
-				it("Should send correct params to deleteItem if we pass in an entire object with unnecessary attributes with range key", async () => {
+        it("Should send correct params to deleteItem if we pass in an entire object with unnecessary attributes with range key", async () => {
```
Comment (fishcharlie) [inline test/unit/Model.js:3705]:
```suggestion
				it("Should send correct params to deleteItem if we pass in an entire object with unnecessary attributes with range key", async () => {
```

File: test/unit/Model.js [hunk 10]
```diff
@@ -3720,6 +3720,24 @@ describe("Model", () => {
...
+				it("Should successfully add a condition if a condition is passed in", async () => {
+ 					const condition = new dynamoose.Condition().filter("id").exists();
+ 					deleteItemFunction = () => Promise.resolve();
+ 					await callType.func(User).bind(User)({"id": 1}, {"condition": condition});
+ 					expect(deleteItemParams).to.be.an("object");
+ 					expect(deleteItemParams).to.eql({
+ 						"Key": {
+ 							"id": {
+ 								"N": "1"
+ 							}
```
Comment (fishcharlie) [inline test/unit/Model.js:3732]:
```suggestion
				it("Should successfully add a condition if a condition is passed in", async () => {
					const condition = new dynamoose.Condition().filter("id").exists();
					deleteItemFunction = () => Promise.resolve();
					await callType.func(User).bind(User)({"id": 1}, {"condition": condition});
					expect(deleteItemParams).to.be.an("object");
					expect(deleteItemParams).to.eql({
						"Key": {
							"id": {
								"N": "1"
 							}
```

File: test/unit/Model.js [hunk 11]
```diff
@@ -3720,6 +3720,24 @@ describe("Model", () => {
...
+								"N": "1"
+ 							}
+ 						},
+ 						"TableName": "User",
+ 						"ConditionExpression": "attribute_exists (#a0)",
+ 						"ExpressionAttributeNames": {
+ 							"#a0": "id"
+ 						}
+ 					});
+ 				});
```
Comment (fishcharlie) [inline test/unit/Model.js:3740]:
```suggestion
							}
						},
						"TableName": "User",
						"ConditionExpression": "attribute_exists (#a0)",
						"ExpressionAttributeNames": {
							"#a0": "id"
						}
					});
				});
```

**Pull request merged**